File: 0676_Implement Magic Dictionary/ImplementMagicDictionary.py
import collections
class MagicDictionary(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
    def buildDict(self, words):
        """
        Build a dictionary through a list of words
        :type dict: List[str]
        :rtype: None
        """
        self.dict = collections.defaultdict(set)
        for word in words:
            for i in xrange(len(word)):
                self.dict[word[:i] + '*' + word[i+1:]].add(word[i])
    
    def search(self, word):
        """
        Returns if there is any word in the trie that equals to the given word after modifying exactly one character
        :type word: str
        :rtype: bool
        """
        for i in xrange(len(word)):
            w = word[:i] + '*' + word[i+1:]
            if w in self.dict and (len(self.dict[w]) > 1 or word[i] not in self.dict[w]): return True        
        return False
            

File: 0676_Implement Magic Dictionary/ImplementMagicDictionary.cs
 public class MagicDictionary
    {
        private Dictionary<string, HashSet<char>> dictionary;
        /** Initialize your data structure here. */
        public MagicDictionary()
        {
            dictionary = new Dictionary<string, HashSet<char>>();
        }
 
        /** Build a dictionary through a list of words */
        public void BuildDict(string[] dict)
        {
            var sb = new StringBuilder();
 
            foreach (var str in dict)   
            {
                for(int i=0;i<str.Length;i++)
                {
                    sb.Clear();
                    sb.Append(str);
                    sb[i] = '*';
                    var newstr = sb.ToString();
 
                    if (!dictionary.ContainsKey(newstr))
                    {
                        dictionary.Add(newstr, new HashSet<char>());
                    }
                    
                    dictionary[newstr].Add(str[i]);
                }
            }
        }
 
        /** Returns if there is any word in the trie that equals to the given word after modifying exactly one character */
        public bool Search(string word)
        {
            var sb = new StringBuilder();
            for (int i = 0; i < word.Length; i++)
            {
                sb.Clear();
                sb.Append(word);
                sb[i] = '*';
                var newstr = sb.ToString();
                if (dictionary.ContainsKey(newstr))
                {
                    var set = dictionary[newstr];
                    if (set.Count > 1) return true;
                    if (!set.Contains(word[i])) return true;
                }
            }
 
            return false;
        }
    }
 
    /**
     * Your MagicDictionary object will be instantiated and called as such:
     * MagicDictionary obj = new MagicDictionary();
     * obj.BuildDict(dict);
     * bool param_2 = obj.Search(word);
     */
